# Exploration

In [3]:
import pandas as pd

# Load the dataset
# We use low_memory=False because PrimeKG has mixed types in some columns 
# and this prevents a common pandas warning.
df = pd.read_csv('kg.csv', low_memory=False)

# Display the first 5 rows and the total number of rows/columns
print(f"Dataset loaded with {df.shape[0]} rows and {df.shape[1]} columns.")
df.head(n=100)

Dataset loaded with 8100498 rows and 12 columns.


relation display_relation  x_index  x_id        x_type  x_name  \
0   protein_protein              ppi        0  9796  gene/protein  PHYHIP   
1   protein_protein              ppi        1  7918  gene/protein  GPANK1   
2   protein_protein              ppi        2  8233  gene/protein   ZRSR2   
3   protein_protein              ppi        3  4899  gene/protein    NRF1   
4   protein_protein              ppi        4  5297  gene/protein   PI4KA   
..              ...              ...      ...   ...           ...     ...   
95  protein_protein              ppi       92  1203  gene/protein    CLN5   
96  protein_protein              ppi       93  1847  gene/protein   DUSP5   
97  protein_protein              ppi       94  9921  gene/protein   RNF10   
98  protein_protein              ppi       95  4726  gene/protein  NDUFS6   
99  protein_protein              ppi       51  6667  gene/protein     SP1   

   x_source  y_index    y_id        y_type  y_name y_source  
0      NCBI     8889   56992  gene/protein   KIF15     NCBI  
1      NCBI     2798    9240  gene/protein   PNMA1     NCBI  
2      NCBI     5646   23548  gene/protein   TTC33     NCBI  
3      NCBI    11592   11253  gene/protein  MAN1B1     NCBI  
4      NCBI     2122    8601  gene/protein   RGS20     NCBI  
..      ...      ...     ...           ...     ...      ...  
95     NCBI     6190  203562  gene/protein  TMEM31     NCBI  
96     NCBI      285    1994  gene/protein  ELAVL1     NCBI  
97     NCBI     7757   79003  gene/protein   MIS12     NCBI  
98     NCBI     8204   23530  gene/protein     NNT     NCBI  
99     NCBI     2761    6874  gene/protein    TAF4     NCBI  

[100 rows x 12 columns]

In [4]:
# Randomly sample 10 rows from the dataset
random_sample = df.sample(n=10, random_state=42)
random_sample

relation         display_relation  x_index  \
6014191          disease_protein          associated with    27390   
2663834                drug_drug  synergistic interaction    17646   
5240480  anatomy_protein_present       expression present    78502   
2228718                drug_drug  synergistic interaction    15337   
5609802          protein_protein                      ppi     1887   
2815777                drug_drug  synergistic interaction    21032   
2301515                drug_drug  synergistic interaction    14061   
2693463                drug_drug  synergistic interaction    14840   
5295073  anatomy_protein_present       expression present    61177   
4684510  anatomy_protein_present       expression present      286   

                                                      x_id        x_type  \
6014191  13183_14026_14614_13450_32811_14872_9758_10241...       disease   
2663834                                            DB00565          drug   
5240480                                             440456  gene/protein   
2228718                                            DB01558          drug   
5609802                                               9531  gene/protein   
2815777                                            DB15360          drug   
2301515                                            DB00586          drug   
2693463                                            DB00392          drug   
5295073                                          101340251  gene/protein   
4684510                                               4929  gene/protein   

                                        x_name       x_source  y_index  \
6014191  congenital stationary night blindness  MONDO_grouped     4757   
2663834                          Cisatracurium       DrugBank    19160   
5240480                              PLEKHM1P1           NCBI    64427   
2228718                             Bromazepam       DrugBank    20324   
5609802                                   BAG3           NCBI      119   
2815777                      Carfentanil, C-11       DrugBank    16287   
2301515                             Diclofenac       DrugBank    16878   
2693463                            Profenamine       DrugBank    17396   
5295073                               SNORD124           NCBI    64487   
4684510                                  NR4A2           NCBI    63180   

            y_id        y_type                y_name  y_source  
6014191     2784  gene/protein                  GNB3      NCBI  
2663834  DB13520          drug           Metergoline  DrugBank  
5240480     1987       anatomy              placenta    UBERON  
2228718  DB06824          drug  Triethylenetetramine  DrugBank  
5609802     2113  gene/protein                  ETS1      NCBI  
2815777  DB04917          drug            Renzapride  DrugBank  
2301515  DB11699          drug           Tropisetron  DrugBank  
2693463  DB01336          drug            Metocurine  DrugBank  
5295073     2048       anatomy                  lung    UBERON  
4684510       81       anatomy           metanephros    UBERON

In [5]:
# Count the occurrences of each human-readable relation
edge_counts = df['display_relation'].value_counts()

# Convert to a dataframe for a cleaner look
edge_info = edge_counts.reset_index()
edge_info.columns = ['Relationship Type', 'Count']

# Display the entire list
print(f"There are {len(edge_info)} unique types of relationships in PrimeKG:")
edge_info

There are 18 unique types of relationships in PrimeKG:


Relationship Type    Count
0        expression present  3036406
1   synergistic interaction  2672628
2            interacts with   686550
3                       ppi   642150
4         phenotype present   300634
5              parent-child   281744
6           associated with   167482
7               side effect   129568
8          contraindication    61350
9         expression absent    39774
10                   target    32760
11               indication    18776
12                   enzyme    10634
13              transporter     6184
14            off-label use     5136
15                linked to     4608
16         phenotype absent     2386
17                  carrier     1728

In [6]:
# We look at both X and Y types because some types might only appear on one side
# Combine both columns and count unique values
node_types = pd.concat([df['x_type'], df['y_type']]).value_counts()

# Create a clean summary table
node_info = node_types.reset_index()
node_info.columns = ['Entity Type', 'Total Occurrences']

print(f"There are {len(node_info)} unique categories of entities in PrimeKG:")
node_info

There are 10 unique categories of entities in PrimeKG:


Entity Type  Total Occurrences
0                drug            5611392
1        gene/protein            5262458
2             anatomy            3132308
3             disease             682488
4    effect/phenotype             514192
5  biological_process             504404
6  molecular_function             193446
7  cellular_component             186204
8             pathway              95432
9            exposure              18672

In [7]:
# Create a cross-tabulation of source types and target types
metagraph = pd.crosstab(df['x_type'], df['y_type'])

# Display the table
# We use a heatmap-style view if you're in a notebook to make it easier to read
metagraph

y_type              anatomy  biological_process  cellular_component  disease  \
x_type                                                                         
anatomy               28064                   0                   0        0   
biological_process        0              105772                   0        0   
cellular_component        0                   0                9690        0   
disease                   0                   0                   0    64388   
drug                      0                   0                   0    42631   
effect/phenotype          0                   0                   0   151510   
exposure                  0                1625                  10     2304   
gene/protein        1538090              144805               83402    80411   
molecular_function        0                   0                   0        0   
pathway                   0                   0                   0        0   

y_type                 drug  effect/phenotype  exposure  gene/protein  \
x_type                                                                  
anatomy                   0                 0         0       1538090   
biological_process        0                 0      1625        144805   
cellular_component        0                 0        10         83402   
disease               42631            151510      2304         80411   
drug                2672628             64784         0         25653   
effect/phenotype      64784             37472         0          3330   
exposure                  0                 0      4140          1212   
gene/protein          25653              3330      1212        642150   
molecular_function        0                 0        45         69530   
pathway                   0                 0         0         42646   

y_type              molecular_function  pathway  
x_type                                           
anatomy                              0        0  
biological_process                   0        0  
cellular_component                   0        0  
disease                              0        0  
drug                                 0        0  
effect/phenotype                     0        0  
exposure                            45        0  
gene/protein                     69530    42646  
molecular_function               27148        0  
pathway                              0     5070

In [8]:
# Hub nodes

# Combine X and Y names to see which entities appear most often across the whole graph
all_nodes = pd.concat([df['x_name'], df['y_name']])
top_nodes = all_nodes.value_counts().head(15)

# Create a clean summary
top_nodes_df = top_nodes.reset_index()
top_nodes_df.columns = ['Entity Name', 'Number of Connections (Edges)']

print("The 15 most connected 'Hub' nodes in the dataset:")
top_nodes_df

The 15 most connected 'Hub' nodes in the dataset:


Entity Name  Number of Connections (Edges)
0          multi-cellular organism                          34710
1                  small intestine                          33710
2                           testis                          33486
3                   fallopian tube                          33484
4                   prostate gland                          33238
5                           spleen                          33200
6                        intestine                          33084
7   dorsolateral prefrontal cortex                          33026
8                        esophagus                          32952
9                          stomach                          32946
10                   thyroid gland                          32912
11          adult mammalian kidney                          32840
12                   adrenal gland                          32418
13                           colon                          32284
14          saliva-secreting gland                          32096

In [10]:
# Define the VIP types for drug repurposing
repurposing_types = ['drug', 'disease', 'gene/protein']

# Filter the dataframe to only include rows where BOTH X and Y are in our VIP list
repurposing_df = df[
    df['x_type'].isin(repurposing_types) & 
    df['y_type'].isin(repurposing_types)
]

# Now let's see the top hubs within THIS filtered group
top_repurposing_hubs = pd.concat([repurposing_df['x_name'], repurposing_df['y_name']]).value_counts().head(15)

print("Top hubs specifically for Drug-Disease-Gene interactions:")
top_repurposing_hubs

Top hubs specifically for Drug-Disease-Gene interactions:


UBC               10396
Chlorpromazine     5070
Quinidine          5026
Amitriptyline      4954
Clozapine          4870
Imipramine         4826
Haloperidol        4682
Doxepin            4668
Carbamazepine      4638
Clomipramine       4628
Zonisamide         4562
Quetiapine         4550
Duloxetine         4520
Desipramine        4472
Paliperidone       4468
Name: count, dtype: int64

In [11]:
# Create a filter for the 'Repurposing Triangle'
# We want to see how these 3 types interact with each other
triplet_relations = repurposing_df['display_relation'].value_counts()

print("The most common 'Verbs' linking Drugs, Diseases, and Proteins:")
triplet_relations.head(15)

The most common 'Verbs' linking Drugs, Diseases, and Proteins:


display_relation
synergistic interaction    2672628
ppi                         642150
associated with             160822
parent-child                 64388
contraindication             61350
target                       32760
indication                   18776
enzyme                       10634
transporter                   6184
off-label use                 5136
carrier                       1728
Name: count, dtype: int64

In [13]:
# 1. Filter the dataset for Aspirin (search in both x and y just in case)
aspirin_edges = df[(df['x_name'].str.lower() == 'aspirin') | (df['y_name'].str.lower() == 'aspirin')]

# 2. See what types of nodes Aspirin is shaking hands with
neighborhood_summary = aspirin_edges.apply(
    lambda row: row['y_type'] if row['x_name'].lower() == 'aspirin' else row['x_type'], 
    axis=1
).value_counts()

print("Aspirin's Neighborhood Composition:")
print(neighborhood_summary)

Aspirin's Neighborhood Composition:
disease    2
Name: count, dtype: int64


In [14]:
# 1. See what those 2 diseases actually are
print("The 2 direct disease connections found:")
print(aspirin_edges[['x_name', 'display_relation', 'y_name']])

# 2. Check if 'Aspirin' exists under other names or capitalization
all_names = pd.concat([df['x_name'], df['y_name']]).unique()
potential_matches = [name for name in all_names if 'aspirin' in str(name).lower()]

print("\nOther potential names for Aspirin in the dataset:")
print(potential_matches)

The 2 direct disease connections found:
                  x_name display_relation           y_name
3783741          Aspirin        linked to  prostate cancer
6497579  prostate cancer        linked to          Aspirin

Other potential names for Aspirin in the dataset:
['Nitroaspirin', 'Carbaspirin calcium', 'asthma, nasal polyps, and aspirin intolerance', 'Aspirin', 'Exacerbated by aspirin ingestion', 'Aspirin-induced asthma', 'aspirin resistance', 'Biosynthesis of aspirin-triggered D-series resolvins']


In [19]:
# Filter for the 'bogus' Aspirin node again
bogus_aspirin = df[df['x_name'] == 'Aspirin']

# Let's look at the specific IDs and sources for this node
print("Metadata for the 'Aspirin' node:")
bogus_aspirin[['x_index', 'x_id', 'x_type', 'x_source']].drop_duplicates()

Metadata for the 'Aspirin' node:


x_index     x_id    x_type x_source
3783741    61900  D001241  exposure      CTD

In [15]:
# Search for 'aspirin' specifically where the entity type is 'drug'
true_drug_aspirin = df[
    ((df['x_name'].str.lower() == 'aspirin') & (df['x_type'] == 'drug')) | 
    ((df['y_name'].str.lower() == 'aspirin') & (df['y_type'] == 'drug'))
]

# If that's empty, we search for its chemical name: 'Acetylsalicylic acid'
chemical_name_aspirin = df[
    (df['x_name'].str.contains('Acetylsalicylic', case=False)) | 
    (df['y_name'].str.contains('Acetylsalicylic', case=False))
]

print(f"Found {len(true_drug_aspirin)} rows for 'Aspirin' (Type: Drug)")
print(f"Found {len(chemical_name_aspirin)} rows for 'Acetylsalicylic acid'")

Found 0 rows for 'Aspirin' (Type: Drug)
Found 3956 rows for 'Acetylsalicylic acid'


In [16]:
# 1. Filter using the correct name
real_aspirin_df = df[(df['x_name'] == 'Acetylsalicylic acid') | (df['y_name'] == 'Acetylsalicylic acid')]

# 2. Get the composition of entity types it connects to
real_summary = real_aspirin_df.apply(
    lambda row: row['y_type'] if row['x_name'] == 'Acetylsalicylic acid' else row['x_type'], 
    axis=1
).value_counts()

print("The REAL Aspirin (Acetylsalicylic acid) Neighborhood:")
print(real_summary)

The REAL Aspirin (Acetylsalicylic acid) Neighborhood:
drug                3784
effect/phenotype      76
gene/protein          70
disease               26
Name: count, dtype: int64


In [20]:
# Filter for rows where Aspirin (the drug) connects to a gene/protein
aspirin_targets = real_aspirin_df[
    (real_aspirin_df['x_type'] == 'gene/protein') | 
    (real_aspirin_df['y_type'] == 'gene/protein')
]

# Get the names of these genes
# We check both x and y name depending on which side the gene is on
target_names = aspirin_targets.apply(
    lambda row: row['y_name'] if row['x_name'] == 'Acetylsalicylic acid' else row['x_name'], 
    axis=1
).unique()

print(f"Aspirin's {len(target_names)} Molecular Gateways:")
print(target_names)

Aspirin's 35 Molecular Gateways:
['CYP2C9' 'CYP2C19' 'NAT2' 'UGT1A6' 'PTGS1' 'AKR1C1' 'PTGS2' 'CASP1'
 'EDNRA' 'CCND1' 'HSPA5' 'IKBKB' 'CASP3' 'TP53' 'NFKBIA' 'MYC' 'RPS6KA3'
 'PRKAA1' 'PRKAA2' 'PRKAB1' 'PRKAB2' 'PRKAG1' 'PRKAG2' 'PRKAG3' 'TNFAIP6'
 'PCNA' 'MAPK1' 'MAPK15' 'MAPK3' 'MAPK4' 'MAPK6' 'MAPK7' 'ABCB1' 'SLC22A6'
 'SLC22A8']


In [21]:
# 1. Get the degree (total connections) for every node in the whole graph
all_nodes = pd.concat([df['x_name'], df['y_name']])
node_counts = all_nodes.value_counts()

# 2. Map those counts back to our 35 Aspirin gateways
gateway_stats = pd.DataFrame({
    'Gateway Name': target_names,
    'Degree (Total Connections)': [node_counts.get(name, 0) for name in target_names]
}).sort_values(by='Degree (Total Connections)', ascending=False)

print("Which gateways are the 'Loudest' (Highest Degree)?")
gateway_stats.head(10)

Which gateways are the 'Loudest' (Highest Degree)?


Gateway Name  Degree (Total Connections)
15          MYC                        2874
13         TP53                        2670
26        MAPK1                        1688
32        ABCB1                        1532
28        MAPK3                        1380
10        HSPA5                        1216
25         PCNA                        1080
6         PTGS2                        1056
0        CYP2C9                        1010
9         CCND1                         976

Notice that PTGS2 (COX-2) is on your Top 10 list with 1,056 connections, but PTGS1 (the primary target for Aspirin's anti-platelet/heart effect) didn't even make the Top 10.

    The Bias: A naive algorithm will naturally favor a path through PTGS2 or MAPK1 simply because they offer more "branches" to explore.

    The Discovery: To find the real mechanism for Aspirin → Myocardial Infarction, an algorithm has to be "smart" enough to ignore the 2,800 connections of MYC and prioritize the (likely) much smaller neighborhood of PTGS1



If we go through Aspirin → TP53 → Disease we get a low score (short cut), and if we go through Aspirin → PTGS1 → Disease we get a high score (mechanistic)

In [23]:
# Check for any direct edge between the drug and the target disease
direct_links = df[
    ((df['x_name'] == 'Acetylsalicylic acid') & (df['y_name'].str.lower() == 'myocardial infarction')) |
    ((df['y_name'] == 'Acetylsalicylic acid') & (df['x_name'].str.lower() == 'myocardial infarction'))
]

print(f"Found {len(direct_links)} direct (1-hop) edges.")
direct_links[['x_name', 'display_relation', 'y_name']]

Found 2 direct (1-hop) edges.


x_name display_relation                 y_name
366906    Acetylsalicylic acid       indication  myocardial infarction
5753706  myocardial infarction       indication   Acetylsalicylic acid

In [22]:
# 1. Define our target disease (using the common name in PrimeKG)
target_disease = 'myocardial infarction'

# 2. Find edges where any of our 35 genes connect directly to the disease
shortcuts = df[
    ((df['x_name'].isin(target_names)) & (df['y_name'].str.lower() == target_disease)) |
    ((df['y_name'].isin(target_names)) & (df['x_name'].str.lower() == target_disease))
]

# 3. Create a summary of these "Shortcut Gates"
shortcut_summary = shortcuts.apply(
    lambda row: row['x_name'] if row['y_name'].lower() == target_disease else row['y_name'],
    axis=1
).value_counts().reset_index()

shortcut_summary.columns = ['Gene Gate', 'Connection Count']
# Add the degree we calculated earlier to see if these are Hubs
shortcut_summary['Gate_Total_Degree'] = shortcut_summary['Gene Gate'].map(node_counts)

print(f"Found {len(shortcut_summary)} potential 2-hop shortcuts to Myocardial Infarction:")
shortcut_summary

Found 2 potential 2-hop shortcuts to Myocardial Infarction:


Gene Gate  Connection Count  Gate_Total_Degree
0     CASP3                 2                806
1     HSPA5                 2               1216

In [24]:
# Look at the neighbors of the 'True' target, PTGS1
ptgs1_neighbors = df[(df['x_name'] == 'PTGS1') | (df['y_name'] == 'PTGS1')]

# Summarize what types of biology PTGS1 is involved in
ptgs1_summary = ptgs1_neighbors.apply(
    lambda row: row['y_type'] if row['x_name'] == 'PTGS1' else row['x_type'], 
    axis=1
).value_counts()

print("PTGS1 (The Correct Gate) Neighborhood:")
print(ptgs1_summary)

PTGS1 (The Correct Gate) Neighborhood:
drug                  226
anatomy               224
disease                54
biological_process     16
molecular_function     12
cellular_component     12
gene/protein           10
effect/phenotype        4
pathway                 4
exposure                2
Name: count, dtype: int64


In [25]:
# 1. Get the names of the Biological Processes and Pathways connected to PTGS1
ptgs1_biology = ptgs1_neighbors[
    (ptgs1_neighbors['x_type'].isin(['biological_process', 'pathway'])) | 
    (ptgs1_neighbors['y_type'].isin(['biological_process', 'pathway']))
]

ptgs1_bio_names = ptgs1_biology.apply(
    lambda row: row['y_name'] if row['x_name'] == 'PTGS1' else row['x_name'], 
    axis=1
).unique()

# 2. Check if any of THESE connect to Myocardial Infarction
mechanistic_bridges = df[
    ((df['x_name'].isin(ptgs1_bio_names)) & (df['y_name'].str.lower() == 'myocardial infarction')) |
    ((df['y_name'].isin(ptgs1_bio_names)) & (df['x_name'].str.lower() == 'myocardial infarction'))
]

print(f"Found {len(mechanistic_bridges)} biological bridges from PTGS1 to the disease:")
mechanistic_bridges[['x_name', 'display_relation', 'y_name']]

Found 0 biological bridges from PTGS1 to the disease:


Empty DataFrame
Columns: [x_name, display_relation, y_name]
Index: []

In [27]:
# 1. Identify the 16 processes correctly
ptgs1_biology = ptgs1_neighbors[
    (ptgs1_neighbors['x_type'].isin(['biological_process', 'pathway'])) | 
    (ptgs1_neighbors['y_type'].isin(['biological_process', 'pathway']))
]

# Get names safely
ptgs1_bio_names = []
for _, row in ptgs1_biology.iterrows():
    name = row['y_name'] if row['x_name'] == 'PTGS1' else row['x_name']
    ptgs1_bio_names.append(name)

# 2. Check the neighbors of THESE processes to find the bridge to MI
# We use .str.lower() safely on the dataframe column here
bridge_to_mi = df[
    ((df['x_name'].isin(ptgs1_bio_names)) & (df['y_name'].str.lower() == 'myocardial infarction')) |
    ((df['y_name'].isin(ptgs1_bio_names)) & (df['x_name'].str.lower() == 'myocardial infarction'))
]

print("The 16 Biological Processes/Pathways connected to PTGS1:")
print(ptgs1_bio_names)

print(f"\nDirect connections from these processes to Myocardial Infarction: {len(bridge_to_mi)}")
if len(bridge_to_mi) > 0:
    print(bridge_to_mi[['x_name', 'display_relation', 'y_name']])

The 16 Biological Processes/Pathways connected to PTGS1:
['inflammatory response', 'response to oxidative stress', 'regulation of blood pressure', 'regulation of cell population proliferation', 'cellular oxidant detoxification', 'cyclooxygenase pathway', 'prostaglandin biosynthetic process', 'prostanoid metabolic process', 'COX reactions', 'Synthesis of Prostaglandins (PG) and Thromboxanes (TX)', 'inflammatory response', 'response to oxidative stress', 'regulation of blood pressure', 'regulation of cell population proliferation', 'cellular oxidant detoxification', 'cyclooxygenase pathway', 'prostaglandin biosynthetic process', 'prostanoid metabolic process', 'COX reactions', 'Synthesis of Prostaglandins (PG) and Thromboxanes (TX)']

Direct connections from these processes to Myocardial Infarction: 0


In [28]:
# 1. Target the specific "Thromboxane" pathway node
target_pathway = 'Synthesis of Prostaglandins (PG) and Thromboxanes (TX)'

# 2. Get its neighbors
pathway_neighbors = df[(df['x_name'] == target_pathway) | (df['y_name'] == target_pathway)]

# 3. Identify what it connects to
pathway_bridges = pathway_neighbors.apply(
    lambda row: row['y_name'] if row['x_name'] == target_pathway else row['x_name'], 
    axis=1
).unique()

# 4. Check if any of THESE neighbors connect to Myocardial Infarction
deep_bridge = df[
    ((df['x_name'].isin(pathway_bridges)) & (df['y_name'].str.lower() == 'myocardial infarction')) |
    ((df['y_name'].isin(pathway_bridges)) & (df['x_name'].str.lower() == 'myocardial infarction'))
]

print(f"Neighbors of the Thromboxane Pathway: {len(pathway_bridges)}")
print(pathway_bridges[:15]) # Show first 15

print(f"\nDo any of these neighbors finally touch the disease? {len(deep_bridge)}")
if len(deep_bridge) > 0:
    print(deep_bridge[['x_name', 'display_relation', 'y_name']])

Neighbors of the Thromboxane Pathway: 16
['Arachidonic acid metabolism' 'PTGES3' 'PRXL2B' 'PTGR2' 'CYP8B1' 'HPGDS'
 'HPGD' 'PTGDS' 'PTGIS' 'PTGS1' 'PTGS2' 'TBXAS1' 'PTGES2' 'AKR1C3' 'CBR1']

Do any of these neighbors finally touch the disease? 0


In [30]:
def find_mechanistic_path(df, start_node, target_disease, max_hops=10):
    # Normalize target for matching
    target_disease = target_disease.lower()
    
    # queue stores (current_node, current_path_list)
    queue = [(start_node, [start_node])]
    visited = {start_node}
    
    print(f"Starting discovery from {start_node}...")

    for hop in range(1, max_hops + 1):
        next_queue = []
        print(f"Exploring Hop {hop}... (Queue size: {len(queue)})")
        
        # Get all neighbors for the current frontier
        current_nodes = [node for node, path in queue]
        neighbors_df = df[df['x_name'].isin(current_nodes) | df['y_name'].isin(current_nodes)]
        
        for _, row in neighbors_df.iterrows():
            # Determine which end of the edge is the "new" neighbor
            # We need to find which node in the row is NOT the one we are currently expanding
            # This is a bit tricky in a vectorized way, so we'll iterate through the frontier
            pass # (Logic handled below for clarity)

        # To keep it efficient, we process the frontier
        for current_node, path in queue:
            # Find all neighbors of the current_node
            edges = df[(df['x_name'] == current_node) | (df['y_name'] == current_node)]
            
            for _, row in edges.iterrows():
                neighbor = row['y_name'] if row['x_name'] == current_node else row['x_name']
                
                if neighbor not in visited:
                    new_path = path + [neighbor]
                    
                    # Check if we hit the finish line
                    if neighbor.lower() == target_disease:
                        print(f"\nSUCCESS! Found path at hop {hop}:")
                        return " -> ".join(new_path)
                    
                    visited.add(neighbor)
                    next_queue.append((neighbor, new_path))
        
        queue = next_queue
        if not queue:
            break
            
    return "No path found within hop limit."

# Execution
# We start from PTGS1 because we've already established Aspirin -> PTGS1
ground_truth = find_mechanistic_path(df, 'PTGS1', 'myocardial infarction', max_hops=10)
print(ground_truth)

Starting discovery from PTGS1...
Exploring Hop 1... (Queue size: 1)
Exploring Hop 2... (Queue size: 280)

SUCCESS! Found path at hop 2:
PTGS1 -> Eletriptan -> myocardial infarction


In [31]:
# 1. We know these are the 'Official' intermediate processes from our previous cell
official_processes = [
    'Synthesis of Prostaglandins (PG) and Thromboxanes (TX)',
    'cyclooxygenase pathway',
    'prostaglandin biosynthetic process'
]

# 2. Find all neighbors of these processes (looking for the 4th hop)
next_leg = df[df['x_name'].isin(official_processes) | df['y_name'].isin(official_processes)]

# 3. Identify these '4th hop' nodes
hop4_nodes = []
for _, row in next_leg.iterrows():
    neighbor = row['y_name'] if row['x_name'] in official_processes else row['x_name']
    hop4_nodes.append(neighbor)
hop4_nodes = list(set(hop4_nodes))

# 4. NOW, search from these hop4_nodes to the disease
# This is where we see if the official path 'connects' to the finish line
final_check = df[
    ((df['x_name'].isin(hop4_nodes)) & (df['y_name'].str.lower() == 'myocardial infarction')) |
    ((df['y_name'].isin(hop4_nodes)) & (df['x_name'].str.lower() == 'myocardial infarction'))
]

print(f"Nodes in the 4th hop of the mechanistic chain: {len(hop4_nodes)}")
print(f"Connections to Myocardial Infarction: {len(final_check)}")

if len(final_check) > 0:
    print("\nGROUND TRUTH FOUND! The bridge node(s) is:")
    print(final_check[['x_name', 'display_relation', 'y_name']])
else:
    print("\nStill no connection. The 'official' chain might be broken in PrimeKG.")

Nodes in the 4th hop of the mechanistic chain: 28
Connections to Myocardial Infarction: 0

Still no connection. The 'official' chain might be broken in PrimeKG.


In [32]:
# 1. Start from our 'Official' Thromboxane processes
official_processes = [
    'Synthesis of Prostaglandins (PG) and Thromboxanes (TX)',
    'cyclooxygenase pathway',
    'prostaglandin biosynthetic process'
]

# 2. Get neighbors of these processes (looking for the 4th hop)
hop4_neighbors = df[df['x_name'].isin(official_processes) | df['y_name'].isin(official_processes)]

# 3. Filter for nodes that are NOT drugs or high-degree hubs
# This forces the search to stay in 'Biology'
hop4_nodes = []
for _, row in hop4_neighbors.iterrows():
    neighbor = row['y_name'] if row['x_name'] in official_processes else row['x_name']
    n_type = row['y_type'] if row['x_name'] in official_processes else row['x_type']
    
    # We want biological things, not other drugs
    if n_type in ['gene/protein', 'biological_process', 'effect/phenotype', 'pathway']:
        hop4_nodes.append(neighbor)

hop4_nodes = list(set(hop4_nodes))

# 4. Final check: Do any of these biological nodes connect to the disease?
final_check = df[
    ((df['x_name'].isin(hop4_nodes)) & (df['y_name'].str.lower() == 'myocardial infarction')) |
    ((df['y_name'].isin(hop4_nodes)) & (df['x_name'].str.lower() == 'myocardial infarction'))
]

print(f"Biological nodes in the 4th hop: {len(hop4_nodes)}")
if len(final_check) > 0:
    print("GROUND TRUTH BIOLOGICAL BRIDGE FOUND:")
    print(final_check[['x_name', 'display_relation', 'y_name']])
else:
    print("The official biological chain is still missing a link to the disease.")

Biological nodes in the 4th hop: 28
The official biological chain is still missing a link to the disease.


In [33]:
# 1. Get the neighbors of our 28 biological nodes (Hop 5 from PTGS1)
hop5_neighbors_df = df[df['x_name'].isin(hop4_nodes) | df['y_name'].isin(hop4_nodes)]

hop5_nodes = set()
for _, row in hop5_neighbors_df.iterrows():
    neighbor = row['y_name'] if row['x_name'] in hop4_nodes else row['x_name']
    hop5_nodes.add(neighbor)

# 2. Get the neighbors of Myocardial Infarction (The disease's "neighborhood")
disease_neighbors_df = df[
    (df['x_name'].str.lower() == 'myocardial infarction') | 
    (df['y_name'].str.lower() == 'myocardial infarction')
]

disease_neighborhood = set()
for _, row in disease_neighbors_df.iterrows():
    neighbor = row['y_name'] if row['x_name'].lower() == 'myocardial infarction' else row['x_name']
    disease_neighborhood.add(neighbor)

# 3. Find the collision
collision = hop5_nodes.intersection(disease_neighborhood)

print(f"Nodes 5 hops away from PTGS1: {len(hop5_nodes)}")
print(f"Nodes 1 hop away from MI: {len(disease_neighborhood)}")
print(f"\nFOUND THE MISSING LINK(S): {collision}")

Nodes 5 hops away from PTGS1: 1733
Nodes 1 hop away from MI: 329

FOUND THE MISSING LINK(S): {'Etodolac', 'Eletriptan', 'Acetylsalicylic acid', 'NOS2', 'Meclofenamic acid', 'Tolmetin', 'Tretinoin', 'Ketorolac', 'Capsaicin', 'Piroxicam', 'ESR1', 'SOD1', 'Drospirenone', 'BAX', 'Meloxicam', 'CREB1', 'Oxaprozin', 'Naproxen', 'Rosiglitazone', 'Diclofenac', 'Minoxidil', 'Indomethacin', 'CASP3', 'NR3C2', 'CKB', 'Diflunisal', 'Flurbiprofen', 'Ketamine', 'Fenoprofen', 'Epirubicin', 'Ketoprofen', 'Sulindac', 'Nabumetone', 'Ibuprofen', 'Celecoxib', 'Mefenamic acid'}


In [35]:
def fast_mechanistic_search(df, start_node, target_node, node_degrees, max_degree=500):
    # 1. Setup
    target_node = target_node.lower()
    forward_visited = {start_node: [start_node]}
    backward_visited = {target_node: [target_node]}
    
    forward_queue = [start_node]
    backward_queue = [target_node]
    
    for hop in range(1, 5): # We shouldn't need more than 4-5 hops for a specific path
        print(f"Expanding Hop {hop}...")
        
        # --- Forward Step ---
        next_forward = []
        for current in forward_queue:
            neighbors = df[(df['x_name'] == current) | (df['y_name'] == current)]
            for _, row in neighbors.iterrows():
                neighbor = row['y_name'] if row['x_name'] == current else row['x_name']
                
                # Filter: Ignore Hubs and stay away from "Shortcut" categories (other drugs)
                deg = node_degrees.get(neighbor, 0)
                n_type = row['y_type'] if row['x_name'] == current else row['x_type']
                
                if deg < max_degree and n_type != 'drug':
                    if neighbor not in forward_visited:
                        forward_visited[neighbor] = forward_visited[current] + [neighbor]
                        next_forward.append(neighbor)
                        
                        if neighbor in backward_visited:
                            print("!!! COLLISION FOUND FROM FORWARD STEP !!!")
                            return forward_visited[neighbor] + backward_visited[neighbor][::-1][1:]

        forward_queue = next_forward
        
        # --- Backward Step ---
        next_backward = []
        for current in backward_queue:
            # Note: handle the lower() case for the disease start
            neighbors = df[(df['x_name'].str.lower() == current.lower()) | (df['y_name'].str.lower() == current.lower())]
            for _, row in neighbors.iterrows():
                neighbor = row['y_name'] if row['x_name'].lower() == current.lower() else row['x_name']
                
                deg = node_degrees.get(neighbor, 0)
                if deg < max_degree:
                    if neighbor not in backward_visited:
                        backward_visited[neighbor] = backward_visited[current] + [neighbor]
                        next_backward.append(neighbor)
                        
                        if neighbor in forward_visited:
                            print("!!! COLLISION FOUND FROM BACKWARD STEP !!!")
                            return forward_visited[neighbor] + backward_visited[neighbor][::-1][1:]
        backward_queue = next_backward
        
    return "No specific mechanistic path found."

# Run it
path = fast_mechanistic_search(df, 'Acetylsalicylic acid', 'myocardial infarction', node_counts)
print("\nResulting Path:")
print(path)

Expanding Hop 1...
Expanding Hop 2...
!!! COLLISION FOUND FROM FORWARD STEP !!!

Resulting Path:
['Acetylsalicylic acid', 'migraine with or without aura, susceptibility to', 'PHACTR1', 'myocardial infarction']


The Hub Trap: Jumps through high-degree proteins (CASP3).

The Drug Trap: Jumps through other drugs (Eletriptan).

The Disease/Trait Trap: Jumps through unrelated diseases that share genetic risk factors (PHACTR1).

In [36]:
def pure_mechanistic_search(df, start_node, target_node, node_degrees, max_degree=500):
    target_node = target_node.lower()
    forward_visited = {start_node: [start_node]}
    backward_visited = {target_node: [target_node]}
    
    forward_queue = [start_node]
    backward_queue = [target_node]
    
    for hop in range(1, 6): # Increased to 5 hops to allow for the longer mechanistic path
        print(f"Expanding Hop {hop}...")
        
        # Forward Step
        next_forward = []
        for current in forward_queue:
            neighbors = df[(df['x_name'] == current) | (df['y_name'] == current)]
            for _, row in neighbors.iterrows():
                neighbor = row['y_name'] if row['x_name'] == current else row['x_name']
                n_type = row['y_type'] if row['x_name'] == current else row['x_type']
                deg = node_degrees.get(neighbor, 0)
                
                # CONSTRAINT: No 'disease' nodes in the middle of the path
                if deg < max_degree and n_type != 'disease' and n_type != 'drug':
                    if neighbor not in forward_visited:
                        forward_visited[neighbor] = forward_visited[current] + [neighbor]
                        next_forward.append(neighbor)
                        if neighbor in backward_visited:
                            return forward_visited[neighbor] + backward_visited[neighbor][::-1][1:]
        forward_queue = next_forward
        
        # Backward Step
        next_backward = []
        for current in backward_queue:
            # The disease itself is the starting point for backward, so we allow it there
            neighbors = df[(df['x_name'].str.lower() == current.lower()) | (df['y_name'].str.lower() == current.lower())]
            for _, row in neighbors.iterrows():
                neighbor = row['y_name'] if row['x_name'].lower() == current.lower() else row['x_name']
                n_type = row['y_type'] if row['x_name'].lower() == current.lower() else row['x_type']
                deg = node_degrees.get(neighbor, 0)
                
                if deg < max_degree and (n_type != 'disease' or neighbor.lower() == target_node):
                    if neighbor not in backward_visited:
                        backward_visited[neighbor] = backward_visited[current] + [neighbor]
                        next_backward.append(neighbor)
                        if neighbor in forward_visited:
                            return forward_visited[neighbor] + backward_visited[neighbor][::-1][1:]
        backward_queue = next_backward
        
    return "No pure biological path found within constraints."

# Run the 'Pure' search
pure_path = pure_mechanistic_search(df, 'Acetylsalicylic acid', 'myocardial infarction', node_counts)
print("\nPure Mechanistic Path:")
print(pure_path)

Expanding Hop 1...
Expanding Hop 2...

Pure Mechanistic Path:
['Acetylsalicylic acid', 'SLC22A6', 'Riboflavin', 'myocardial infarction']


In [37]:
# Another failure

# The project is to "Benchmark the failure of current algorithms"


def calculate_path_quality(path, df, node_degrees):
    """
    Scores the 'Biological Plausibility' of a path.
    """
    if not path or isinstance(path, str): return 0
    
    score = 0
    # 1. Specificity Bonus: Paths through 'Biological Process' or 'Pathway' are gold
    # 2. Hub Penalty: Deduct points for high-degree nodes
    # 3. Relay Bonus: Points for hitting the 'Thromboxane' node specifically
    
    for i, node in enumerate(path):
        deg = node_degrees.get(node, 0)
        # Penalize Hubs (Non-specific noise)
        if deg > 500: score -= 20 
        
        # Reward specific biology
        if "thromboxane" in node.lower() or "prostaglandin" in node.lower():
            score += 50 
            
    return score

# --- The Comparison Experiment ---
results = []

# 1. The Naive Approach (Dijkstra/Shortest)
# Result: ['Aspirin', 'Eletriptan', 'MI']
naive_path = ['Acetylsalicylic acid', 'Eletriptan', 'myocardial infarction'] 

# 2. The Semantic Shortcut (PHACTR1)
# Result: ['Aspirin', 'Migraine', 'PHACTR1', 'MI']
genetic_path = ['Acetylsalicylic acid', 'migraine with or without aura, susceptibility to', 'PHACTR1', 'myocardial infarction']

# 3. Your 'Mechanistic' Attempt (The best we found)
# Note: Even if it didn't reach MI, we score the route it took
mechanistic_route = ['Acetylsalicylic acid', 'PTGS1', 'Synthesis of Prostaglandins (PG) and Thromboxanes (TX)']

# --- Generate Scores ---
for name, p in [("Naive", naive_path), ("Genetic", genetic_path), ("Mechanistic", mechanistic_route)]:
    q_score = calculate_path_quality(p, df, node_counts)
    results.append({"Method": name, "Path": " -> ".join(p), "Quality Score": q_score})

import pandas as pd
benchmark_df = pd.DataFrame(results)
print(benchmark_df)

        Method                                               Path  \
0        Naive  Acetylsalicylic acid -> Eletriptan -> myocardi...   
1      Genetic  Acetylsalicylic acid -> migraine with or witho...   
2  Mechanistic  Acetylsalicylic acid -> PTGS1 -> Synthesis of ...   

   Quality Score  
0            -60  
1            -40  
2             10  


#### Warfarin check


In [38]:
# 1. Who does Warfarin talk to?
warfarin_neighbors = df[
    (df['x_name'].str.lower() == 'warfarin') | 
    (df['y_name'].str.lower() == 'warfarin')
]

print(f"Warfarin has {len(warfarin_neighbors)} direct connections.")
print("\nDirect Neighbors (First 10):")
print(warfarin_neighbors[['x_name', 'display_relation', 'y_name']].head(10))

# 2. Specifically check for the 'Vitamin K' bridge
vit_k_check = df[
    (df['x_name'].str.contains('Vitamin K', case=False)) | 
    (df['y_name'].str.contains('Vitamin K', case=False))
]
print(f"\nVitamin K related nodes found: {len(vit_k_check)}")

Warfarin has 4296 direct connections.

Direct Neighbors (First 10):
          x_name display_relation   y_name
321294  Warfarin          carrier      ALB
321710  Warfarin          carrier     ORM1
323541  Warfarin           enzyme   CYP1A2
323985  Warfarin           enzyme   CYP3A4
324875  Warfarin           enzyme   CYP2C9
325214  Warfarin           enzyme   CYP2C8
325744  Warfarin           enzyme  CYP2C19
326553  Warfarin           enzyme  CYP2C18
332895  Warfarin           target   VKORC1
334685  Warfarin           target    NR1I2

Vitamin K related nodes found: 310
